libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h

```diff
@@ -0,0 +1,197 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA__MEMORY_RESOURCE_CUDA_MEMORY_RESOURCE_H
+#define _CUDA__MEMORY_RESOURCE_CUDA_MEMORY_RESOURCE_H
+
+#include <cuda/__cccl_config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda_runtime_api.h>
+
+#include <cuda/__memory_resource/get_property.h>
+#include <cuda/__memory_resource/properties.h>
+#include <cuda/__memory_resource/resource_ref.h>
+#include <cuda/__memory_resource/resource.h>
+#include <cuda/stream_ref>
+
+#if _CCCL_STD_VER >= 2014
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA_MR
+
+/**
+ * @brief `cuda_memory_resource` uses cudaMalloc/Free for allocation/deallocation.
+ */
+struct cuda_memory_resource
+{
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation. Is ignored!
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(const size_t __bytes, size_t) const
+  {
+    return allocate(__bytes);
+  }
```

Comment:
cudaMalloc is guaranteed to return memory that is aligned to at least 256B. Instead of ignoring the alignment, we should check that it is <= 256, otherwise throw.

Reply:
I went with a simple assert until we are sure what we want to do regarding exceptions

Reply:
What remains to be decided about exceptions? We're already throwing exceptions in the `stream_ref` code: https://github.com/NVIDIA/cccl/blob/d88fbd7341647533128f2bbe65cea9f22a874caf/libcudacxx/include/cuda/stream_ref#L139-L154

Reply:
Actually only me needing to write a proper wrapper for that